### ISSUE TITLE
[BUG] get_sdist_metadata return value in tests can change

### ISSUE THREAD

**Issue by woutdenolf** [conda-incubator/grayskull#361]
```
(py38) denolf@lindenolf:~/dev/grayskull$ pip list | grep souschef
conda-souschef                2.2.1
(py38) denolf@lindenolf:~/dev/grayskull$ pip list | grep gray
grayskull                     1.5.1.dev3+g62509c4.d20220728      (current main branch)
```

This test passes when I run it like this

```
pytest tests/test_pypi.py::test_compose_test_section_with_console_scripts
```

but fails when I do this

```
pytest tests/test_pypi.py
```

Somehow `get_sdist_metadata` returns something different in both cases. Not sure why.

**Comment by marcelotrevisani** [comment]
I can't reproduce it, both ways work for me.
here is my ``conda list``
```bash
❯ conda list                                             
# packages in environment at /home/bb8/mambaforge/envs/gs-dev:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
alabaster                 0.7.12                     py_0    conda-forge
attrs                     21.4.0             pyhd8ed1ab_0    conda-forge
babel                     2.10.3             pyhd8ed1ab_0    conda-forge
binutils                  2.36.1               hdd6e379_2    conda-forge
binutils_impl_linux-64    2.36.1               h193b22a_2    conda-forge
binutils_linux-64         2.36                hf3e587d_10    conda-forge
black                     22.6.0             pyhd8ed1ab_0    conda-forge
brotlipy                  0.7.0           py310h5764c6d_1004    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
c-compiler                1.4.2                h166bdaf_0    conda-forge
ca-certificates           2022.6.15            ha878542_0    conda-forge
certifi                   2022.6.15       py310hff52083_0    conda-forge
cffi                      1.15.1          py310h255011f_0    conda-forge
cfgv                      3.3.1              pyhd8ed1ab_0    conda-forge
charset-normalizer        2.1.0              pyhd8ed1ab_0    conda-forge
click                     8.1.3           py310hff52083_0    conda-forge
colorama                  0.4.5              pyhd8ed1ab_0    conda-forge
compilers                 1.4.2                ha770c72_0    conda-forge
conda-souschef            2.2.1              pyhd8ed1ab_0    conda-forge
coverage                  6.4.2           py310h5764c6d_0    conda-forge
cryptography              37.0.4          py310h597c629_0    conda-forge
cxx-compiler              1.4.2                h924138e_0    conda-forge
dataclasses               0.8                pyhc8e2a94_3    conda-forge
distlib                   0.3.5              pyhd8ed1ab_0    conda-forge
docutils                  0.18.1          py310hff52083_1    conda-forge
execnet                   1.9.0              pyhd8ed1ab_0    conda-forge
filelock                  3.7.1              pyhd8ed1ab_0    conda-forge
fortran-compiler          1.4.2                h2a4ca65_0    conda-forge
gcc                       10.3.0              he2824d0_10    conda-forge
gcc_impl_linux-64         10.3.0              hf2f2afa_16    conda-forge
gcc_linux-64              10.3.0              hc39de41_10    conda-forge
gfortran                  10.3.0              h18518b4_10    conda-forge
gfortran_impl_linux-64    10.3.0              h73f4979_16    conda-forge
gfortran_linux-64         10.3.0              hb09a455_10    conda-forge
gxx                       10.3.0              he2824d0_10    conda-forge
gxx_impl_linux-64         10.3.0              hf2f2afa_16    conda-forge
gxx_linux-64              10.3.0              h2593f52_10    conda-forge
identify                  2.5.2              pyhd8ed1ab_0    conda-forge
idna                      3.3                pyhd8ed1ab_0    conda-forge
imagesize                 1.4.1              pyhd8ed1ab_0    conda-forge
importlib-metadata        4.11.4          py310hff52083_0    conda-forge
iniconfig                 1.1.1              pyh9f0ad1d_0    conda-forge
jarowinkler               1.2.0           py310hd8f1fbe_0    conda-forge
jinja2                    3.1.2              pyhd8ed1ab_1    conda-forge
kernel-headers_linux-64   2.6.32              he073ed8_15    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
libblas                   3.9.0           15_linux64_openblas    conda-forge
libcblas                  3.9.0           15_linux64_openblas    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-devel_linux-64     10.3.0              he6cfe16_16    conda-forge
libgcc-ng                 12.1.0              h8d9b700_16    conda-forge
libgfortran-ng            12.1.0              h69a702a_16    conda-forge
libgfortran5              12.1.0              hdcd56e2_16    conda-forge
libgomp                   12.1.0              h8d9b700_16    conda-forge
liblapack                 3.9.0           15_linux64_openblas    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libopenblas               0.3.20          pthreads_h78a6416_0    conda-forge
libsanitizer              10.3.0              h26c7422_16    conda-forge
libstdcxx-devel_linux-64  10.3.0              he6cfe16_16    conda-forge
libstdcxx-ng              12.1.0              ha89aaad_16    conda-forge
libuuid                   2.32.1            h7f98852_1000    conda-forge
libzlib                   1.2.12               h166bdaf_2    conda-forge
markupsafe                2.1.1           py310h5764c6d_1    conda-forge
mock                      4.0.3           py310hff52083_3    conda-forge
mypy_extensions           0.4.3           py310hff52083_5    conda-forge
ncurses                   6.3                  h27087fc_1    conda-forge
nodeenv                   1.7.0              pyhd8ed1ab_0    conda-forge
numpy                     1.23.1          py310h53a5b5f_0    conda-forge
openssl                   1.1.1q               h166bdaf_0    conda-forge
packaging                 21.3               pyhd8ed1ab_0    conda-forge
pathspec                  0.9.0              pyhd8ed1ab_0    conda-forge
pip                       22.1.2             pyhd8ed1ab_0    conda-forge
pkginfo                   1.8.3              pyhd8ed1ab_0    conda-forge
platformdirs              2.5.2              pyhd8ed1ab_1    conda-forge
pluggy                    1.0.0           py310hff52083_3    conda-forge
pre-commit                2.20.0          py310hff52083_0    conda-forge
progressbar2              4.0.0              pyhd8ed1ab_0    conda-forge
py                        1.11.0             pyh6c4a22f_0    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pygments                  2.12.0             pyhd8ed1ab_0    conda-forge
pyopenssl                 22.0.0             pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.9              pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1           py310hff52083_5    conda-forge
pytest                    7.1.2           py310hff52083_0    conda-forge
pytest-console-scripts    1.3.1              pyhd8ed1ab_0    conda-forge
pytest-cov                3.0.0              pyhd8ed1ab_0    conda-forge
pytest-forked             1.4.0              pyhd8ed1ab_0    conda-forge
pytest-mock               3.8.2              pyhd8ed1ab_0    conda-forge
pytest-xdist              2.5.0              pyhd8ed1ab_0    conda-forge
python                    3.10.5          h582c2e5_0_cpython    conda-forge
python-utils              3.3.3              pyhd8ed1ab_0    conda-forge
python_abi                3.10                    2_cp310    conda-forge
pytz                      2022.1             pyhd8ed1ab_0    conda-forge
pyyaml                    6.0             py310h5764c6d_4    conda-forge
rapidfuzz                 2.2.0           py310hd8f1fbe_0    conda-forge
readline                  8.1.2                h0f457ee_0    conda-forge
requests                  2.28.1             pyhd8ed1ab_0    conda-forge
ruamel.yaml               0.17.21         py310h5764c6d_1    conda-forge
ruamel.yaml.clib          0.2.6           py310h5764c6d_1    conda-forge
ruamel.yaml.jinja2        0.2.4                      py_1    conda-forge
setuptools                63.2.0          py310hff52083_0    conda-forge
setuptools-scm            7.0.5              pyhd8ed1ab_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
snowballstemmer           2.2.0              pyhd8ed1ab_0    conda-forge
sphinx                    5.0.2              pyh6c4a22f_0    conda-forge
sphinxcontrib-applehelp   1.0.2                      py_0    conda-forge
sphinxcontrib-devhelp     1.0.2                      py_0    conda-forge
sphinxcontrib-htmlhelp    2.0.0              pyhd8ed1ab_0    conda-forge
sphinxcontrib-jsmath      1.0.1                      py_0    conda-forge
sphinxcontrib-qthelp      1.0.3                      py_0    conda-forge
sphinxcontrib-serializinghtml 1.1.5              pyhd8ed1ab_2    conda-forge
sqlite                    3.39.1               h4ff8645_0    conda-forge
stdlib-list               0.7.0                      py_2    conda-forge
sysroot_linux-64          2.12                he073ed8_15    conda-forge
tk                        8.6.12               h27826a3_0    conda-forge
toml                      0.10.2             pyhd8ed1ab_0    conda-forge
tomli                     2.0.1              pyhd8ed1ab_0    conda-forge
tomli-w                   1.0.0              pyhd8ed1ab_0    conda-forge
tomlkit                   0.11.1             pyha770c72_0    conda-forge
typed-ast                 1.5.4           py310h5764c6d_0    conda-forge
typing                    3.10.0.0           pyhd8ed1ab_0    conda-forge
typing-extensions         4.3.0                hd8ed1ab_0    conda-forge
typing_extensions         4.3.0              pyha770c72_0    conda-forge
tzdata                    2022a                h191b570_0    conda-forge
ukkonen                   1.0.1           py310hbf28c38_2    conda-forge
urllib3                   1.26.10            pyhd8ed1ab_0    conda-forge
virtualenv                20.15.1         py310hff52083_0    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
yaml                      0.2.5                h7f98852_2    conda-forge
zipp                      3.8.0              pyhd8ed1ab_0    conda-forge
zlib                      1.2.12               h166bdaf_2    conda-forge
```

![image](https://user-images.githubusercontent.com/7882269/181712678-b4ec1e04-589e-4a26-b9fd-fd3fcc5f57d4.png)

**Comment by woutdenolf** [comment]
Some debugging led me to `injection_distutils`.

It seems this method is absorbing any exception (comment: never absorb `BaseException`, but that's another discussion)

```python
@contextmanager
def injection_distutils(folder: str, debug=False) -> dict:
     ...
    try:
        __run_setup_py(path_setup, data_dist)
        ...
    except BaseException as err:
        raise
        log.debug(f"Exception occurred when executing sdist injection: {err}")
        yield data_dist
    finally:
        core.setup = setup_core_original
        os.chdir(old_dir)
```

The exception we get when running `pytest tests/test_pypi.py` is this

```python
_____________________________________________________________________________________ test_compose_test_section_with_console_scripts _____________________________________________________________________________________

    def test_compose_test_section_with_console_scripts():
        config = Configuration(name="pytest", version="7.1.2")
        metadata1 = get_pypi_metadata(config)
>       metadata2 = get_sdist_metadata(
            "https://pypi.io/packages/source/p/pytest/pytest-7.1.2.tar.gz", config, debug=True
        )

tests/test_pypi.py:247: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
grayskull/strategy/py_base.py:694: in get_sdist_metadata
    with injection_distutils(temp_folder, debug=debug) as metadata:
/users/denolf/.pyenv/ubuntu_20_04/versions/3.8.5/lib/python3.8/contextlib.py:113: in __enter__
    return next(self.gen)
grayskull/strategy/py_base.py:385: in injection_distutils
    __run_setup_py(path_setup, data_dist)
grayskull/strategy/py_base.py:442: in __run_setup_py
    merge_sdist_metadata(data_dist, get_setup_cfg(os.path.dirname(str(path_setup))))
grayskull/strategy/py_base.py:289: in get_setup_cfg
    setup_cfg = read_configuration(str(path_setup_cfg))
/users/denolf/virtualenvs/ewoks/ubuntu_20_04/py38/lib/python3.8/site-packages/setuptools/config.py:87: in read_configuration
    dist = Distribution()
/users/denolf/virtualenvs/ewoks/ubuntu_20_04/py38/lib/python3.8/site-packages/setuptools/dist.py:459: in __init__
    _Distribution.__init__(
/users/denolf/.pyenv/ubuntu_20_04/versions/3.8.5/lib/python3.8/distutils/dist.py:292: in __init__
    self.finalize_options()
/users/denolf/virtualenvs/ewoks/ubuntu_20_04/py38/lib/python3.8/site-packages/setuptools/dist.py:837: in finalize_options
    ep(self)
/tmp/grayskull-pytest-d060832v/pytest-5.3.1/.eggs/setuptools_scm-7.0.5-py3.8.egg/setuptools_scm/integration.py:123: in infer_version
    _assign_version(dist, config)
/tmp/grayskull-pytest-d060832v/pytest-5.3.1/.eggs/setuptools_scm-7.0.5-py3.8.egg/setuptools_scm/integration.py:59: in _assign_version
    maybe_version = _get_version(config)
/tmp/grayskull-pytest-d060832v/pytest-5.3.1/.eggs/setuptools_scm-7.0.5-py3.8.egg/setuptools_scm/__init__.py:154: in _get_version
    parsed_version = _do_parse(config)
/tmp/grayskull-pytest-d060832v/pytest-5.3.1/.eggs/setuptools_scm-7.0.5-py3.8.egg/setuptools_scm/__init__.py:101: in _do_parse
    version = _version_from_entrypoints(config) or _version_from_entrypoints(
/tmp/grayskull-pytest-d060832v/pytest-5.3.1/.eggs/setuptools_scm-7.0.5-py3.8.egg/setuptools_scm/_entrypoints.py:65: in _version_from_entrypoints
    for ep in iter_matching_entrypoints(root, entrypoint, config):
/tmp/grayskull-pytest-d060832v/pytest-5.3.1/.eggs/setuptools_scm-7.0.5-py3.8.egg/setuptools_scm/discover.py:65: in iter_matching_entrypoints
    for ep in iter_entry_points(entrypoint):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

group = 'setuptools_scm.parse_scm', name = None

    def iter_entry_points(
        group: str, name: str | None = None
    ) -> Iterator[_t.EntrypointProtocol]:
        all_eps = entry_points()
        if hasattr(all_eps, "select"):
            eps = all_eps.select(group=group)
        else:
>           eps = all_eps[group]
E           KeyError: 'setuptools_scm.parse_scm'

/tmp/grayskull-pytest-d060832v/pytest-5.3.1/.eggs/setuptools_scm-7.0.5-py3.8.egg/setuptools_scm/_entrypoints.py:96: KeyError
```

The method `injection_distutils` is quite horrible. There must be a better way. In addition I've seen lots of dodgy stuff concerning `setuptools_scm` all over the code base.

**Comment by marcelotrevisani** [comment]
there are reasons to use ``BaseException`` there, if you look to the code it is logging the stack trace in debug level, so you could just run grayskull with the debugging level enabled and you would see it as well.
The main reason to catch the exception is to gracefully terminate and still generate a generic recipe for the user.

``injection_distutil`` is similar of what conda-build does to recover metadata from distutil, as distutil is quite convoluted to properly recover everything. There was a few discussion in the Python core and python distutils to add ways to recover that data, but last time I checked they don't want to do anything, packaging in Python is quite horrible :)

**Comment by woutdenolf** [comment]
Installing `setuptools-scm` solved the issue. So we need to add `setuptools-scm` to the test requirements of `grayskull`?

**Comment by woutdenolf** [comment]
> The main reason to catch the exception is to gracefully terminate and still generate a generic recipe for the user.

And this cannot be done by catching `Exception` instead of `BaseException`?

**Comment by marcelotrevisani** [comment]
> > The main reason to catch the exception is to gracefully terminate and still generate a generic recipe for the user.
> 
> And this cannot be done by catching `Exception` instead of `BaseException`?

there are a few bugs in the past that was not covered by using just exception. Some exceptions like ``SystemExit`` and ``GeneratorExit`` are not covered by using just ``Exception``, the ``BaseException`` there is needed

**Comment by woutdenolf** [comment]
> there are a few bugs in the past that was not covered by using just exception. Some exceptions like SystemExit and GeneratorExit are not covered by using just Exception, the BaseException there is needed

Ok but this means `grayskull` cannot be interrupted or killed when inside `injection_distutil`.

**Comment by marcelotrevisani** [comment]
> > there are a few bugs in the past that was not covered by using just exception. Some exceptions like SystemExit and GeneratorExit are not covered by using just Exception, the BaseException there is needed
> 
> Ok but this means `grayskull` cannot be interrupted or killed when inside `injection_distutil`.

It cannot be killed by KeyboardInterrupt, but sigkill works as normal

**Comment by marcelotrevisani** [comment]
but it can be improved, indeed, putting something like
```python
except KeyboardInterrupt:
    sys.exit()
```
before the ``BaseException`` to cover this case

**Comment by marcelotrevisani** [comment]
regarding the fix is not necessary to add ``setuptools-scm`` as a dependency, we just need to modify the test. 
On this line:
https://github.com/conda-incubator/grayskull/blob/ad7b2f8ab0b9a143acea5805adb06318d557084c/tests/test_pypi.py#L245

we can change it to be:
```python
    config = Configuration(name="pytest", version="7.1.2", missing_deps={"setuptools-scm"})
```

I tested on my env and I removed ``setuptools-scm`` as a dependency and it worked.

``grayskul`` has a few strategies to install in a isolated env the missing dependencies that are needed in build time.
